Update the Fitting of Peaks in Texture Analysis Pipeline

andy-bridger opened the pull request:
### Description of work

Currently, fitting is somewhat flaky and unreliable, largely owing to two factors: the fact that peaks might be very weak/ nonexistent (i.e. the sample has texture); and the fitting in dspacing doesn't give nice order of magnitude for step size of A/B

To try and address these I've done a bit of an update, firstly fitting in TOF and then converting back to d, and secondly providing better starting parameters and generally improving the fit approach


### To test:

Firstly, ideally you would test the system tests (`TextureAnalysisScriptTest.py`) on mac os

Secondly, func test this script:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
from mantid.api import AnalysisDataService as ADS
from os import path, makedirs, scandir
from Engineering.texture.TextureUtils import find_all_files, fit_all_peaks, mk
from Engineering.common.calibration_info import CalibrationInfo
from Engineering.EnggUtils import GROUP

############### ENGINEERING DIFFRACTION INTERFACE FITTING ANALOGUE #######################

######################### EXPERIMENTAL INFORMATION ########################################

# First, you need to specify your file directories, If you are happy to use the same root, from experiment
# to experiment, you can just change this experiment name.
exp_name = "PostExp-SteelCentre"

# otherwise set root directory here:
root_dir = fr"C:\Users\kcd17618\Engineering_Mantid\User\{exp_name}"

# Next the folder contraining the workspaces you want to fit
file_folder = "Focus"
# These are likely within a sub-folder specified by the detector grouping
grouping = "Texture30"
prm_path = None
groupingfile_path = None

# You also need to specify a name for the folder the fit parameters will be saved in
fit_save_folder = "ScriptFitParameters-FitTest2"

# Provide a list of peaks that you want to be fit within the spectra
peaks = [1.44] # steel
#peaks = [2.8, 2.575, 2.455, 1.89, 1.62, 1.46] # zr

# The fitting has a couple of parameters that deal with when peaks are missing as a result of the texture
# The first parameter is 1_over_sigma_thresh - this determines the minimum value of I/sigma for a fit to be considered as for a valid peak
# any invalid peak will have parameters set to nan by default, but these nans can be overwritten by no_fit_value_dicts and nan_replacement
# no_fit_value_dict takes fitted parameter names and allows you to specify what the unfit value should be eg. {"I":0.0} - if you can't fit intensity
# set the value directly to 0.0
# nan_replacement then happens after this, if a nan_replacement method is given any parameters without an unfit_value provided will have nans replaced
# either with "zeros", or with the min/max/mean value of that parameter (Note: if all the values are nan, the value will remain nan)

i_over_sigma_thresh = 3.0
no_fit_value_dict = {"I": 0.0, "I_est": 0.0}
nan_replacement = "mean"

######################### RUN SCRIPT ########################################

# create output directory
fit_save_dir = path.join(root_dir, fit_save_folder)
mk(fit_save_dir)

# find and load peaks

# get grouping directory name
calib_info = CalibrationInfo(group = GROUP(grouping))
if groupingfile_path:
    calib_info.set_grouping_file(groupingfile_path)
elif prm_path:
    calib_info.set_prm_filepath(prm_path) 
group_folder = calib_info.get_group_suffix()
focussed_data_dir = path.join(root_dir, file_folder, group_folder, "CombinedFiles")
focus_ws_paths = find_all_files(focussed_data_dir)
focus_wss = [path.splitext(path.basename(fp))[0] for fp in focus_ws_paths]
for iws, ws in enumerate(focus_wss):
    if not ADS.doesExist(ws):
        Load(Filename = focus_ws_paths[iws], OutputWorkspace= ws)


# execute the fitting                     
fit_all_peaks(focus_wss, peaks, 0.05, fit_save_dir, i_over_sigma_thresh = i_over_sigma_thresh, 
                nan_replacement = nan_replacement, no_fit_value_dict = no_fit_value_dict, smooth_vals = (3,2),
                tied_bkgs = (False, False), final_fit_raw = False)     









```


<!-- Include sufficient instructions for someone unfamiliar with the application to test.
Ok to refer back to instructions in the issue.
-->

<!-- REMEMBER:
- Add labels, milestones, etc.
- Ensure the base of this PR is correct (e.g. release-next or main)
- Add release notes in separate file as per ([guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)), or justify their absence:
  *This does not require release notes* because <fill in an explanation of why>
-->

<!--  GATEKEEPER: When squashing, remove the section from HERE...  -->
---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

andy-bridger committed:
initial improvements

# Conflicts:
#	scripts/Engineering/texture/TextureUtils.py

# Conflicts:
#	scripts/Engineering/texture/TextureUtils.py

# Conflicts:
#	scripts/Engineering/texture/TextureUtils.py

andy-bridger committed:
rebin for speed

andy-bridger committed:
fix bounds bug

andy-bridger committed:
tidy up code

andy-bridger committed:
allow for tying and fixing

andy-bridger committed:
update validation parameters

andy-bridger committed:
fix unittests

andy-bridger committed:
update poor fit handling

andy-bridger committed:
remove macos guard

andy-bridger committed:
add summed pre-fit

andy-bridger committed:
fix rebunch

andy-bridger committed:
start to tidy up unit tests

andy-bridger committed:
extend unit tests

andy-bridger committed:
update sys test

RichardWaiteSTFC commented on scripts/Engineering/texture/TextureUtils.py:
```diff
@@ -510,9 +649,121 @@ def fit_all_peaks(
+            lower.append(xdat.min())
+            upper.append(xdat.max())
+            step.append(np.diff(xdat).max())
+    return f"{np.max(lower)}, {np.max(step)}, {np.min(upper)}"
```
This is a difficult thing to do in practice, but I have a few thoughts.

1. I think `Rebin` will accept these parameters as a list which might make your life easier!
2. Is it expected that the bin edges of a given spectrum will be different in different workspaces (i.e. can you just use the first workspace)?
3. Have these workspaces already been cropped to the region around the Bragg peak? If not, then you could loose a lot of resolution by taking the maximum bin-width over the range of the data - this will particularly affect low d-spacing peaks at backscattering and perhaps even affect the fit (despite improved stats).
4. If you do want to do this over the whole of the data range it would probably be better to preserve the log- binning rather than use constant bin width
5. I think one could do something similar (perhaps not exactly equivalent) using `extractX` (for behaviour with ragged workspaces see https://github.com/mantidproject/mantid/pull/39932#discussion_r2348616463)
```
lower, upper, bin_width = -np.inf, np.inf, -np.inf
for ws in wss:
    xdat = ws_ceria.extractX()
    lower = max(xdat[xdat>0].min(), lower)
    upper = min(xdat.max(), upper)
    dx = max(np.diff(xdat, axis=1).max(), bin_width)
```
or get `dx` from number of bins in each spectrum or even better actually save maximum `dx/x` for log binning?

A different approach could be to estimate I/sigma for all focussed spectra for a given peak (using my favourite summation...after rough estimate of avg. bg) then fit spectrum with max I/sigma to better constrain parameters? Then you don't need to worry about rebinning?

andy-bridger replied:
I like the idea of keeping the summed spectra rather than just the best individual one (feels like it should give better stats if you have a peak that is frequent but never particularly strong?), but agree on all the other points and have updated to crop around peaks first and then work from there.

RichardWaiteSTFC commented on scripts/Engineering/texture/TextureUtils.py:
```diff
@@ -246,154 +257,230 @@ def validate_abs_corr_inputs(
+        # update constraints around new values
+        intens = max(peak.getParameterValue("I"), 1)
+        x0 = peak.getParameterValue("X0")
+        new_peak.addConstraints(f"{max(intens / 2, 1e-6)}<I<{intens * 2}")
```
I'm not sure the final fit should have a constraint of intensity (and I realise I should check what I do in `IntegratePeaks1DProfile`...). By this point an initial fit, or summation methods should have excluded spectra with insufficient counts. Adding constraints is not ideal, basically should use as few as possible (this is true in general, but particularly true in Mantid due to the implementation which adds a cost penalty). There can always be a final pass through I/sigma looking for spectra to exclude?

RichardWaiteSTFC reviewed (changes requested):
Thanks, results look a lot better and fitting in TOF I think is more numerically stable (though I have also fitted in d). I know there's a rush so I don't want to hold up too much so feel free to push back on any of the comments!

I still have a nagging feeling that it should be possible to use **exactly** the same code as `IntegratePeaks1DProfile` for fitting. I know you have inherited from the relevant class `TexturePeakFunctionGenerator(PeakFunctionGenerator)` - perhaps `PeakFunctiongenerator` could have the common bits refactored out. In any case that's an issue for maintenance!

andy-bridger committed:
initial review changes

andy-bridger committed:
tie params in subsequent fits

andy-bridger committed:
update sys test

andy-bridger committed:
fix unit test

RichardWaiteSTFC reviewed (dismissed):
Functional testing worked I think!
Here are the pole figures I get for the peak at ~2.035 Ang in a few runs (364902, 11, 20, 26,29,35,44) - which I think are aluminium?
<img width="1085" height="385" alt="image" src="https://github.com/user-attachments/assets/03c10646-bd17-48a5-89ee-1cb7f0807b4b" />
Does that look right? 
It might be hard to tell as there are not many points on pole figuee
<img width="380" height="285" alt="image" src="https://github.com/user-attachments/assets/6178e21a-be86-4755-8283-21952a5551c8" />

A couple of thoughts (to be addressed in a different PR):

1. It's difficult to inspect the fits - so difficult to know whether I can unfix any parameters
2. Thinking about it, fitting all spectra from a given run is not ideal - we want to be fitting a given group/spectrum across all runs - such that e.g. A and B are expected to be the same (as these depend on the instrument not the sample) and can be global in the fit. Practically I don't know whether that will work when you don't have that many rotations measured (e.g. at the beginning of an experiment or in a system test). To be discussed!
3. As discussed many times - it would be better to fix A and B in a given group at values extracted form ceria fits! Either by parameterising the resolution/peak profile across the detector, or with a lookup table as per Florencia!

I think we can do B and S well (though S from ceria would be a lower bound on the S on a typical sample). A is a bit more tricky - but possibly not as important?

<img width="1701" height="407" alt="image" src="https://github.com/user-attachments/assets/94d4937f-2d9e-4fb5-b070-266722d33a7d" />

Approving on assumption unit tests pass

RichardWaiteSTFC committed:
Increase tolerance in system tests

Test passes locally on windows, but requires a slightly larger tolerance on Linux.
Also put Mixin classes first in inheritance in class def

RichardWaiteSTFC reviewed (dismissed):
Approved - just loosened the tolerance on 2 system tests (some small differences between my local windows and linux).

cailafinn reviewed (changes requested):
Tested the system test on macOS and seems to be all passing fine. Nice work. 

Just the one comment on a duplicated section that I think got missed:

andy-bridger committed:
remove the now redundant tof error conversion

andy-bridger committed:
fix unit test

cailafinn reviewed (dismissed):
Test still passing, code looks good to me.

andy-bridger committed:
release note
